pkg/pillar/cmd/zfsmanager/handlesnapshotmgr.go

```diff
@@ -0,0 +1,312 @@
+			fmt.Errorf("convert compressratio value failed: %v", err))
+	}
+
+	volSize, err := dataset.GetProperty(libzfs.DatasetPropVolsize)
+	if err != nil {
+		volumeConfig.Errors = append(volumeConfig.Errors,
+			fmt.Errorf("get property volsize for dataset failed %v", err))
+	}
+	volumeConfig.Volsize, err = strconv.ParseUint(volSize.Value, 10, 64)
+	if err != nil {
+		volumeConfig.Errors = append(volumeConfig.Errors,
+			fmt.Errorf("convert Volsize value failed: %v", err))
+	}
+
+	creatinonTime, err := dataset.GetProperty(libzfs.DatasetPropCreation)
+	if err != nil {
+		volumeConfig.Errors = append(volumeConfig.Errors,
+			fmt.Errorf("get property creatinonTime for dataset failed: %v", err))
+	}
+	volumeConfig.CreatinonTime, err = strconv.ParseUint(creatinonTime.Value, 10, 64)
+	if err != nil {
+		volumeConfig.Errors = append(volumeConfig.Errors,
+			fmt.Errorf("convert creatinonTime value failed: %v", err))
+	}
+
+	logicalreferenced, err := dataset.GetProperty(libzfs.DatasetPropLogicalreferenced)
+	if err != nil {
+		volumeConfig.Errors = append(volumeConfig.Errors,
+			fmt.Errorf("get property logicalreferenced for dataset failed: %v", err))
+	}
+	volumeConfig.Logicalreferenced, err = strconv.ParseUint(logicalreferenced.Value, 10, 64)
+	if err != nil {
+		volumeConfig.Errors = append(volumeConfig.Errors,
+			fmt.Errorf("convert logicalreferenced space value failed: %v", err))
+	}
+
+	if snapshotConfig.DisplayName != "" {
+		volumeConfig.DisplayName = snapshotConfig.DisplayName
+	} else {
+		//set default display name
+		displayTimeStr := time.Unix(int64(volumeConfig.CreatinonTime), 0).Format("2006-01-02-15:04:05")
+		volumeConfig.DisplayName = fmt.Sprintf("snapshot-%s", displayTimeStr)
+	}
+	volumeConfig.CmdError = ""
+	volumeConfig.RollbackCounter = snapshotConfig.RollbackCmd.Counter
+
+	return volumeConfig, nil
+}
+
+// getZvolName - returns the full name for the zvol
+func getZvolName(ctx *zfsContext, volumeUUID string) string {
+	zvolList := ctx.subVolumeStatus.GetAll()
+	for _, vc := range zvolList {
+		zvol := vc.(types.VolumeStatus)
+		if volumeUUID == zvol.VolumeID.String() {
+			return zvol.ZVolName()
+		}
+	}
+	return ""
+}
+
+// checkAllSnapshotStatusForVolume - removes all snapshotStatus for snapshots
+// that were deleted by ZFS without any commands for the specified zvol
+func checkAllSnapshotStatusForVolume(ctx *zfsContext, volumeUUID string) {
+	zvolName := getZvolName(ctx, volumeUUID)
+	if zvolName == "" {
+		log.Fatalf("volumeStatus with uuid: %s not found", volumeUUID)
+		return
+	}
+
+	allSnapList := ctx.pubSnapshotStatus.GetAll()
+	for _, vc := range allSnapList {
+		snapshot := vc.(types.ZfsSnapshotStatus)
+		if snapshot.VolumeUUID == volumeUUID {
+			if !zfs.DoesSnapshotExist(zvolName, snapshot.ID) {
+				log.Functionf("Snapshot [%s] with volume UUID [%s] was deleted in ZFS after rollback ops",
+					snapshot.ID, volumeUUID)
+				snapshot.Deleted = true
+				unpublishSnapshotStatus(ctx, snapshot.ID)
```

Comment:
It should be publish. We want to delete status only in case of config remove.

Reply:
Corrected